### PR TITLE
[clap-v3-utils] Define `EncodableKey` and make `keypair_from_path` and `keypair_from_seed` generic functions

### PR DESCRIPTION
**Pull request by samkim-crypto**
#### Problem
The clap utils do not yet support ElGamal keypairs and authenticated encryption key that are needed in cli tools such as the token-cli. Adding support for these types would require creating analogous functions to `keypair_from_path` and `keypair_from_seed` in the clap-v3-utils.

Instead of writing these separate functions, it would be nice to write `keypair_from_path` and `keypair_from_seed` as generic functions that can support any type of cryptographic keys/keypairs that implement a certain interface.

#### Summary of Changes
- Create a trait `EncodableKey` that defines the interface for cryptographic keys/keypairs are read, written, and derived from sources.
- Implement `EncodableKey` for the standard `Keypair` in the sdk.
- Write `keypair_from_path` and `keypair_from_seed` as generic functions using `EncodableKey`.

This is a split PR from https://github.com/solana-labs/solana/pull/30901

Fixes #

### CODE REVIEW

File: sdk/src/signer/mod.rs
Comment (CriesofCarrots) [inline sdk/src/signer/mod.rs:128]: Will these all be relevant for ElGamal keypairs and authenticated encryption keys?

File: clap-v3-utils/src/keypair.rs
Comment (CriesofCarrots) [inline clap-v3-utils/src/keypair.rs:1005]: Will ElGamal keypairs and authenticated encryption keys support all the same Sources that Keypair does?

**Review by CriesofCarrots — COMMENTED** [review]
This seems like a reasonable approach to me. Just a couple questions for my own edification before I approve.

File: sdk/src/signer/mod.rs
Comment (samkim-crypto) [reply sdk/src/signer/mod.rs:128]: Yes these functions are all relevant though for `key_from_seed_and_derivation_path`, I was thinking of outputting a suitable error type like [this](https://github.com/solana-labs/solana/pull/30901/files#diff-7d059813b85f149efcc5236cf7f8c29645b18bb9787cdaa05cec84711b06a8a6R309). It is not that `key_from_seed_and_derivation_path` does not make sense for ElGamal keypairs or authenticated encryption keys, but I am not sure how useful it would be and there is also no crate like `ed25519-dalek-bip32` that can do the parsing for us.

File: clap-v3-utils/src/keypair.rs
Comment (samkim-crypto) [reply clap-v3-utils/src/keypair.rs:1005]: Yeah, I am not sure how many users will actually use some of these sources to derive ElGamal and AE keys, but they should support all sources except for `Usb`. I think this should be fine since `keypair_from_path` (or now `encodable_key_from_path`) errors on `Usb` anyway.

**Review by CriesofCarrots — APPROVED** [review]
Thanks, works for me! Wouldn't hurt to get @t-nelson's ✅  too, esp since he requested the split PRs

File: sdk/src/signer/mod.rs
Comment (t-nelson) [inline sdk/src/signer/mod.rs:115]: could we replace this with an associated type like `type Pubkey: ToString;`

File: sdk/src/signer/mod.rs
Comment (t-nelson) [inline sdk/src/signer/mod.rs:119]: `{read,write}_key_file` can be default implementations here on the trait declaration. all they do is open a file then call `{read,write}_key` respectively

File: sdk/src/signer/mod.rs
Comment (t-nelson) [inline sdk/src/signer/mod.rs:123]: iirc, `DerivationPath` makes some assumptions about being bip32-ed25519/slip0011. does this complicate things?

**Review by t-nelson — COMMENTED** [review]
thanks for splitting this out! got a couple comments

File: sdk/src/signer/mod.rs
Comment (samkim-crypto) [reply sdk/src/signer/mod.rs:115]: That is a good idea! This way, the function will be a lot more general.

File: sdk/src/signer/mod.rs
Comment (samkim-crypto) [reply sdk/src/signer/mod.rs:119]: Agreed!

File: sdk/src/signer/mod.rs
Comment (samkim-crypto) [reply sdk/src/signer/mod.rs:123]: Yeah, ElGamal keys have the same cyclic group structure, so bip32 should still be applicable and same with authenticated encryption key for slip0011. But it would involve sophisticated parsers, so I was thinking about outputting a suitable error type for ElGamal and AE keys like [this](https://github.com/solana-labs/solana/pull/30901/files#diff-7d059813b85f149efcc5236cf7f8c29645b18bb9787cdaa05cec84711b06a8a6R309) (also see [above](https://github.com/solana-labs/solana/pull/30947#discussion_r1151326691)) Maybe we can support it in the future.

File: sdk/src/signer/mod.rs
Comment (samkim-crypto) [reply sdk/src/signer/mod.rs]: Actually, for symmetric keys like `AeKey`, there is no `Pubkey` type. The function `pubkey_string(&self)` can return an error in this case like [here](https://github.com/solana-labs/solana/pull/30901/files#diff-8e0f5e2484afe07494baa6995029cfde17aaa6c3e7030c98f7a0d266790c4a66R222), but if we now have an associated `Pubkey` type in the trait, we have to define a dummy pubkey type for symmetric keys that will never be used, so I am not sure if this is the cleanest way to define things. Wdyt?

Comment (t-nelson) [reply sdk/src/signer/mod.rs]: sounds like maybe this attribute doesn't belong on the trait, then?

Comment (samkim-crypto) [reply sdk/src/signer/mod.rs]: Yeah, it seems like removing it is the way to go.

**Comment by samkim-crypto** [comment]
Added default implementation for `read_key_file` and `write_key_file`. Please take a look!

**Comment by samkim-crypto** [comment]
@t-nelson, I was just wondering if you had some time to look at the updates yet? I think I need an approval to merge :pray:

File: sdk/src/signer/mod.rs
Comment (t-nelson) [reply sdk/src/signer/mod.rs:116]: were you still planning to remove `pubkey_string()`?

Comment (samkim-crypto) [reply sdk/src/signer/mod.rs:116]: I was planning to not have an associated `Pubkey` type, but still have the function `pubkey_string()` because it simplifies [confirming pubkeys](https://github.com/solana-labs/solana/pull/30947/files#diff-541bd0923c958e58e17e227c88be47279e4e25605979f0ad1fc536f63d4ae688R1121) in `encodable_key_from_seed_phrase` and I want `encodable_key_from_seed_phrase` to also support symmetric keys. So when symmetric keys implement `EncodableKey`, I was planning on returning a suitable error type like [this](https://github.com/solana-labs/solana/pull/30901/files#diff-8e0f5e2484afe07494baa6995029cfde17aaa6c3e7030c98f7a0d266790c4a66R222).

The canonical way to make `encodable_key_from_seed_phrase` to work is to probably have two traits, `EncodableKey` (with `pubkey_string` removed) and some general `CryptographicKey` trait that has an associated function `pubkey_string() -> Option<String>` or something like that. Then have `encodable_key_from_seed_phrase<K: EncodableKey + CryptographicKey>(...) -> Result<K, Error>`. But I feel like this is more trouble than it is worth.

File: sdk/src/signer/keypair.rs
Comment (samkim-crypto) [reply sdk/src/signer/keypair.rs:125]: Yep, that is a good idea. I will make those changes.

For the deprecations, how about we do it in a separate PR (so deprecate since `1.16`?)? It would be nice if this current PR gets merged in soon and I am afraid that the deprecations will break CI.

File: sdk/src/signer/mod.rs
Comment (t-nelson) [reply sdk/src/signer/mod.rs:116]: if the confirm thing is all that `pubkey_string` is helping, how about we take it out of `encodable_key_from_path` altogether?  if we want `encodable_key_from_path` to be generic on symmetric keys, even the mention of a pubkey is confusing. would it be sufficient to move the `if confirm_pubkey` block out to a helper and call that on the result of `encodable_key_from_path` optionally from `keypair_from_path` and `keypair_from_seed_phrase`?

File: sdk/src/signer/keypair.rs
Comment (t-nelson) [reply sdk/src/signer/keypair.rs]: yeah fine to add the deprecation attribute separately

File: sdk/src/signer/mod.rs
Comment (samkim-crypto) [reply sdk/src/signer/mod.rs:116]: Oh yes, I don't know how I didn't think of just removing the pubkey part out of the generic helper function. I guess I was too focused on minimizing changes to the existing code. I will remove `pubkey_string`.

**Review by t-nelson — APPROVED** [review]
lgtm now. thanks!

**Pull request merged**
